### PR TITLE
Tests/profile following

### PR DESCRIPTION
**Pull request by devgabrielcoman**
## JIRA
[KWS-2142](https://jira.superawesome.tv/secure/RapidBoard.jspa?rapidView=175&view=detail&selectedIssue=KWS-2142)
[KWS-2312](https://jira.superawesome.tv/secure/RapidBoard.jspa?rapidView=175&view=detail&selectedIssue=KWS-2312)

## Description
Added tests for:
- `UserDataProvider`, `SimpleFeedDataProvider`, `Friends (Following/Followers) DataProvider`
- `UserViewModel`, `SimpleFeedViewModel`, `SimpleFeedItemViewModel`, `FriendsViewModel`

Also fixed some issues that the tests caught! 

Also, the way this testing works, I've indeed created separate `SpyDelegate` classes for different types of delegates used by the data providers, but I've also created a series of Mocks for the following models and services:
- `MockLoggedUser`, `MockConfig`, `MockUser`, `MockImage`, `MockFeedItem`, etc
- `MockUserSocialService`, `MockUserSocialActionsService`, `MockFeedService`

So basically all of the things I **need** to inject into a data provider to make it work, I've mocked. This means I don't have to hit a staging endpoint in order to test things and as long as the Mocks conform to the protocols outlined in `SAProtobufs` then, for all intents and purposes, we run on **real-ish data**.

## Images
N/A

### CODE REVIEW

File: PopJamUI/Classes/LoginView/DataProvider/LoginViewDataProvider.swift [hunk 1]
```diff
@@ -68,13 +68,13 @@ public struct LoginViewDataProvider {
...
-            } else {
-                completion(isValid, !isValid, false)
-            }
+            //let isValid = self.validateCharCount(charCount: password.count, minChar: minChar , maxChar: maxChar!)
+//            if isValid {
+//                let regexPassed = self.validateStringRegex(string: password, regex: regex)
+//                completion(isValid, !regexPassed, regexPassed)
+//            } else {
+//                completion(isValid, !isValid, false)
+//            }
```
Comment (me2k) [inline PopJamUI/Classes/LoginView/DataProvider/LoginViewDataProvider.swift]: Delete the commented code as this is held in source control

File: Example/Tests/Friends/DataProvider/TestFollowingDataProvider_ToggleFollowState.swift [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        provider.toggleFollowState(position: 1);
+        
+        // 4. check the state has been reverted
+        
+        expect(delegate.error).toNot(beNil())
+        expect(delegate.users[1].isFollowed).to(beFalse())
+        expect(delegate.users[1].followersCount).to(equal(5))
+    }
+    
+    public func testMedhodCallWithIndexOutOfBounds () {
```
Comment (Tbruni85) [inline Example/Tests/Friends/DataProvider/TestFollowingDataProvider_ToggleFollowState.swift]: space before ()

File: Example/Tests/Mocks/Models/MockConfig.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+//
+//  Created by Gabriel Coman on 08/04/2018.
+//  Copyright © 2018 CocoaPods. All rights reserved.
+//
+
+import Foundation
+import SAProtobufs
+
+struct MockConfig: SocialConfigModelProtocol {
+    var geo: String = "uk"
```
Comment (Tbruni85) [inline Example/Tests/Mocks/Models/MockConfig.swift]: space missing

File: Example/Tests/Mocks/Models/MockFeedItem.swift [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    var rejamOf: FeedItemModelProtocol? = nil
+    var richContent: RichContentModelProtocol? = nil
+    var stickerPositions: [StickerPositionModelProtocol] = []
+    var subItems: [FeedItemModelProtocol] = []
+    
+    init(id: String, postingUserId: String) {
+        self.id = id
+        self.postingUserId = postingUserId
+    }
+
```
Comment (Tbruni85) [inline Example/Tests/Mocks/Models/MockFeedItem.swift]: space

File: Example/Tests/SimpleFeedView/DataProvider/TestSimpleFeedDataProvider_FetchData.swift [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    
+    override func setUp() {
+        super.setUp()
+    }
+    
+    override func tearDown() {
+        super.tearDown()
+    }
+    
+    public func testMethodCallSuccess () {
```
Comment (Tbruni85) [inline Example/Tests/SimpleFeedView/DataProvider/TestSimpleFeedDataProvider_FetchData.swift]: space before ()

File: Example/Tests/SimpleFeedView/DataProvider/TestSimpleFeedDataProvider_FetchData.swift [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        // 2 - get data twice, no new data should be given
+        
+        provider.fetchData()
+        
+        expect(delegate.items).toNot(beNil())
+        expect(delegate.items.count).to(equal(2))
+        expect(delegate.error).to(beNil())
+    }
+    
+    public func testMethodCallError () {
```
Comment (Tbruni85) [inline Example/Tests/SimpleFeedView/DataProvider/TestSimpleFeedDataProvider_FetchData.swift]: space before ()

File: Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_FetchUserProfile.swift [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+        provider.fetchData();
+
+        // 2. check the delegate would have returned a valid profile and no error
+
+        expect(delegate.user).toNot(beNil())
+        expect(delegate.error).to(beNil())
+    }
+    
+    public func testMethodCallWithError () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_FetchUserProfile.swift]: space before ()

File: Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_FetchUserProfile.swift [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    
+    override func setUp() {
+        super.setUp()
+    }
+    
+    override func tearDown() {
+        super.tearDown()
+    }
+
+    public func testMethodCallWithSuccess () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_FetchUserProfile.swift]: space before ()

File: Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_ToggleFollowState.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    
+    override func setUp() {
+        super.setUp()
+    }
+    
+    override func tearDown() {
+        super.tearDown()
+    }
+    
+    public func testMethodWithSuccess () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_ToggleFollowState.swift]: space before ()

File: Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_ToggleFollowState.swift [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        provider.toggleFollowState();
+
+        // 6. check the state has been reverted
+
+        expect(delegate.error).to(beNil())
+        expect(delegate.user?.isFollowed).to(beFalse())
+        expect(delegate.user?.followersCount).to(equal(5))
+    }
+    
+    public func testMethodWithError () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/DataProvider/TestUserDataProvider_ToggleFollowState.swift]: space before ()

File: Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    
+    override func setUp() {
+        super.setUp()
+    }
+    
+    override func tearDown() {
+        super.tearDown()
+    }
+    
+    public func testViewModelId () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift]: space before ()

File: Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        let user = MockUser(id: "user-id")
+        
+        // when
+        let viewModel = UserViewModel(user: user)
+        
+        // then
+        expect(viewModel.userId).to(equal("user-id"))
+    }
+    
+    public func testViewModelName () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift]: space before ()

File: Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        let user = MockUser(id: "user-id", name: "user-name")
+        
+        // when
+        let viewModel = UserViewModel(user: user)
+        
+        // then
+        expect(viewModel.name).to(equal("user-name"))
+    }
+    
+    public func testViewModelFollowersText () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift]: space before ()

File: Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        let viewModel2 = UserViewModel(user: user2)
+        let viewModel3 = UserViewModel(user: user3)
+        
+        // then
+        expect(viewModel1.numberOfFollowers).to(equal("5"))
+        expect(viewModel2.numberOfFollowers).to(equal("13.3k"))
+        expect(viewModel3.numberOfFollowers).to(equal("1.9M"))
+    }
+    
+    public func testViewModelFollowingText () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift]: space before ()

File: Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        let viewModel2 = UserViewModel(user: user2)
+        let viewModel3 = UserViewModel(user: user3)
+        
+        // then
+        expect(viewModel1.numberOfFollowing).to(equal("5"))
+        expect(viewModel2.numberOfFollowing).to(equal("13.3k"))
+        expect(viewModel3.numberOfFollowing).to(equal("1.9M"))
+    }
+    
+    public func testViewModelFollowStatus () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift]: space before ()

File: Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        let viewModel2 = UserViewModel(user: user2)
+        let viewModel3 = UserViewModel(user: user3)
+        
+        // then
+        expect(viewModel1.followButtonText.lowercased()).to(equal("Follow".lowercased()))
+        expect(viewModel2.followButtonText.lowercased()).to(equal("Following".lowercased()))
+        expect(viewModel3.followButtonText.lowercased()).to(equal("Get following".lowercased()))
+    }
+    
+    public func testViewModelImage () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift]: space before ()

File: Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+        let viewModel2 = UserViewModel(user: user2)
+        let viewModel3 = UserViewModel(user: user3)
+        
+        // then
+        expect(viewModel1.thumbnailImage?.absoluteString).to(equal("https://api.some.com/t/152/image.png"))
+        expect(viewModel2.thumbnailImage?.absoluteString).to(equal("https://api.some.com/sticker.jpg"))
+        expect(viewModel3.thumbnailImage).to(beNil())
+    }
+    
+    public func testViewModelBadge () {
```
Comment (Tbruni85) [inline Example/Tests/UserProfileView/ViewModel/TestUserViewModel.swift]: space before ()

File: PopJamUI/Classes/Friends/Views/FriendsCell.swift [hunk 18]
```diff
@@ -42,89 +42,89 @@ class FriendsCell: UICollectionViewCell {
...
     
     public override init(frame: CGRect) {
         super.init(frame: frame)
-        self.setUpView()
+        setUpView()
     }
     
     // MARK: - Redraw
     
     private func setUpView () {
```
Comment (Tbruni85) [inline PopJamUI/Classes/Friends/Views/FriendsCell.swift]: space before ()

File: PopJamUI/Classes/UserProfileView/Views/UserView.swift [hunk 19]
```diff
@@ -48,12 +48,12 @@ public class UserView: UIView, UserDataProviderDelegate {
...
+        setUpView()
     }
     
     public override init(frame: CGRect) {
         super.init(frame: frame)
-        self.setUpView()
+        setUpView()
     }
     
     public convenience init (frame: CGRect, provider: UserDataProvider) {
```
Comment (Tbruni85) [inline PopJamUI/Classes/UserProfileView/Views/UserView.swift]: space before ()

File: PopJamUI/Classes/UserProfileView/Views/UserView.swift [hunk 20]
```diff
@@ -67,101 +67,101 @@ public class UserView: UIView, UserDataProviderDelegate {
...
+        followersTap = UITapGestureRecognizer(target: self, action: #selector(tapAction))
+        followersStats.addGestureRecognizer(followersTap)
         
-        self.followingTap = UITapGestureRecognizer(target: self, action: #selector(tapAction))
-        self.followingStats.addGestureRecognizer(self.followingTap)
+        followingTap = UITapGestureRecognizer(target: self, action: #selector(tapAction))
+        followingStats.addGestureRecognizer(followingTap)
     }
     
     private func redraw (viewModel: UserViewModel) {
```
Comment (Tbruni85) [inline PopJamUI/Classes/UserProfileView/Views/UserView.swift]: space before ()

File: Example/Tests/Friends/DataProvider/TestFollowingDataProvider_ToggleFollowState.swift (hunk 2 above)
Comment (me2k) [inline Example/Tests/Friends/DataProvider/TestFollowingDataProvider_ToggleFollowState.swift]: Spelling mistake: testMedhodCallWithIndexOutOfBounds Method is spelt wrong
